### ISSUE TITLE
[FR] Shorten (and uniquify) recent file names in `consult-buffer` 

### ISSUE THREAD

**Issue by gusbrs** [minad/consult#713]
I'm just getting acquainted with `consult`, and find `consult-buffer` really nifty. There's one thing in it that I think could be improved though, thus this feature request. Candidates coming from `recentf` can be quite long, since they include the full path of the file (well, almost full, just `abbreviate-file-name`). I probably use longer file names than the average, but it is not difficult to get some larger than the window's width. This brings some inconvenience and problems. Even if it doesn't get actually truncated, it might be more information than (some) people would like to search for. If it does get truncated, the most important information gets literally out of view. Besides, it makes the marginalia annotations be pushed to the side, easily becoming completely hidden. And, once they get beyond the margin, they never come back, not even when only shorter candidates are displayed again.

So, the suggestion would be to shorten these candidates. `file-name-nondirectory` would probably be insufficient here since there might well be too recent files with the same one, but in different directories. So, something in the lines of `uniquify`'s `forward` style would be much welcome. I presume as an option, since I can also see that people might also want the full path as it currently is.

I've tried to look into it, but I couldn't get far. Anything related to text properties or overlays wouldn't work at all, since we want to see what we are searching. On the other hand, playing with `consult--source-recent-file`'s `:item` field, I tried pushing a cons to `items` (so that I'd have an alist of candidates from which I could grab the cdr later on), but it seems a string is expected there (as far as I could tell, at least). So I hit a bumper, I can't see a handle I could grab, hence the request.

PS: A slight side comment on `consult-buffer` which I think deserves mention, but probably not a dedicated request. I think the narrowing key for the "Hidden Buffer" group is quite a premium one, and reserved for the possibly the least used case. I get why it was chosen, but I personally would like to use it for the "Buffer" group, and I did set it like that (with an "h" for the hidden ones, and stole back "b" for bookmarks). However, I noticed that you filter out the space from `consult-narrow-help` and, as a result, the "Buffer" group is not shown in it. How about not filtering `32` out from `consult-narrow-help`, and letting users grab that premium key?

**Comment by minad** [comment]
Thanks for the feedback. Abbreviating recent file names has been proposed before, iirc by @jdtsmith. There are two blockers:

1. Abbreviation must be fast enough for `consult-buffer` even if you have 1000 or more recent files. In particular computing unique file names could potentially be slow.

2. Matching abbreviated files is problematic. One could hide parts of the file names via the `invisible` text property. This would allow you to match on the hidden parts, but the matching would be invisible and non-obvious. Alternatively one removes parts of the file names entirely, but then one loses the ability to match.

For this reason I haven't implemented this. Nevertheless if someone proposes a concrete implementation in the form of a patch, I would be happy to discuss it (No promise that it will get merged).

Regarding the keys, as you probably figured, `SPC` is used because ephemeral buffers start with space and since it matches the behavior of `switch-to-buffer`. Therefore the keys will stay as is. Fortunately it is easy enough to change this in your user configuration if you disagree.

**Comment by minad** [comment]
I should add - you can always experiment and adjust the `:items` function of `consult--source-recent-file`, e.g., such that the file name without the directory is used as item name. There are many examples in the wiki, the README and of course in Consult itself on how to write such sources.

**Comment by oantolin** [comment]
> Fortunately it is easy enough to change this in your user configuration if you disagree.

My reading of the issue is that @gusbrs was happy to change the keys in their personal configuration, and the request was really to change this bit from `consult-narrow-help`:

      (seq-filter (lambda (x) (/= (car x) 32))
                  consult--narrow-keys)

As things are, SPC never gets documented. Maybe it would be better to not filter out SPC but to display it as `SPC` rather than a blank.

**Comment by minad** [comment]
Ah I see, I don't know why SPC is filtered out from consult-narrow-help. I would have to check why I did this. Does something weird happen if you remove the filter? (I never use consult-narrow-help.)

**Comment by oantolin** [comment]
Nothing bad happens, but the help is displayed like `  Hidden Buffer`. It would probably be best to change that to `SPC Hidden Buffer`.

**Comment by oantolin** [comment]
I think that's a good solution.

**Comment by minad** [comment]
Yes, I assume this was some leftover from back then when SPC had more special treatment. As far as I know the special treatment has been removed when I added the `:hidden` field to sources.

**Comment by gusbrs** [comment]
> * Abbreviation must be fast enough for `consult-buffer` even if you have 1000 or more recent files. In particular computing unique file names could potentially be slow.

Of course uniquifying has a computational cost. But the default value for `recentf-max-saved-items` is 20. I can see people wanting larger numbers (I'm one myself), but isn't 1000 extreme? Besides, I suggested an option, opt-in, as it were. And I also enumerated a number of drawbacks the long candidates cause. Don't you think the computational cost could be outweighed by not having those problems? And couldn't each user decide about it?
 
> * Matching abbreviated files is problematic. One could hide parts of the file names via the `invisible` text property. This would allow you to match on the hidden parts, but the matching would be invisible and non-obvious. Alternatively one removes parts of the file names entirely, but then one loses the ability to match.

That's why I excluded text properties as a way to go about this, as mentioned in the OP. But some people may be happy to match against `file-name-nondirectory` (just uniquified when needed). I certainly do. I don't have any "preference statistics" on this, but that's actually `ivy-switch-buffer`'s default (not uniquified though, one loses the duplicates). So I think it is safe to presume I'm not alone.

> For this reason I haven't implemented this. Nevertheless if someone proposes a concrete implementation in the form of a patch, I would be happy to experiment with it and discuss it (No promise that it will get merged).

True that I'm new to the package, but I've tried to tackle it, with scant success. I probably have some learning to to before I could offer something of the sort.

> My reading of the issue is that @gusbrs was happy to change the keys in their personal configuration, and the request was really to change this bit from `consult-narrow-help`:
> 
> ```
>   (seq-filter (lambda (x) (/= (car x) 32))
>               consult--narrow-keys)
> ```

Yes, that's what I meant. I've seen the commit, thank you!

> I should add - you can always experiment and adjust the `:items` function of `consult--source-recent-file`, e.g., such that the file name without the directory is used as item name. There are many examples in the wiki, the README and of course in Consult itself on how to write such sources.

I have actually tried that, but since I could only ever pass a list of strings to `:item` (not an alist), just the `file-name-nondirectory` loses the path information in a non-recoverable way, and that bit is not sufficient to take action on it. But, it is probably my fault. (I did read all the manuals and wikis, but that doesn't mean all the info has sunk in...).

**Comment by jdtsmith** [comment]
See also minad/vertico#198.

**Comment by minad** [comment]
@jdtsmith Yes, truncating the candidates in the UI would always be an option. One could a `metadata` extension `truncate-candidate` which takes values `left`, `right` or `nil`.

Then I just had an another interesting idea. One could introduce special hidden fields in the candidate strings.

~~~ elisp
(concat
 "Some Title"
 (propertize " author:Author " 'invisible t 'completion-invisible-field 'author)
 (propertize " journal:Journal " 'invisible t 'completion-invisible-field 'journal))
~~~

As soon as the completion style matches a `completion-invisible-field`., the field is uncovered. I believe this could be useful for the recent file abbreviation issue discussed here and also packages like Citar. 

One problem is that this needs special support either in the completion UI or in the completion styles. It could be implemented in the UI by comparing the original candidate and the highlighted candidate. If the face property changed of a `completion-invisible-field`, the field would be uncovered. Another problem is that the uncovering messes with the careful candidate formatting used by Citar. cc @bdarcus

**Comment by minad** [comment]
@jdtsmith Instead of advising `vertico--format-candidate` you could also use the transformation facilities described in the Vertico wiki: https://github.com/minad/vertico/wiki#alternative-1-using-vertico-multiform-recommended. I also considered before to add official `vertico-transformation-functions` to Vertico, but haven't found enough use cases.

**Comment by minad** [comment]
> The official response for searching metadata is to embark-export then use e.g. consult-focus-lines. That's fine, but complicated enough that I rarely reach for it.

Yes, but you have to consider that computing annotations can be very expensive. Therefore it is a good design to compute those lazily and not make them available for matching.

> But maybe vertico/marginalia could use this trick to (optionally) put the same annotation data in both the candidate and the annotation.

Not possible because of performance.

**Comment by gusbrs** [comment]
I'm not sure it is of interest, but it might be for someone (@jdtsmith possibly), so I might as well share. Thanks to @minad (at https://old.reddit.com/r/emacs/comments/10arjk1/shortening_candidates_in_consultbuffer/), I was able to find the handle to pass the information around, and came up with a couple of possible approaches.

The first one, simple and cheap, is just using `file-name-nondirectory`. The disadvantage is that candidates with a repeated nondirectory name get shadowed in the completion. But at least the most recent one gets listed. Depending on the use case, it may be a valid option.

```elisp
(defun gb/consult--source-recentf-items ()
  (let ((ht (consult--buffer-file-hash))
        file-name-handler-alist ;; No Tramp slowdown please.
        items)
    (dolist (file recentf-list (nreverse items))
      ;; Emacs 29 abbreviates file paths by default, see
      ;; `recentf-filename-handlers'.
      (unless (eq (aref file 0) ?/)
        (setq file (expand-file-name file)))
      (unless (gethash file ht)
        (push (propertize
               (file-name-nondirectory file)
               'multi-category `(file . ,file))
              items)))))

(plist-put consult--source-recent-file
           :items #'gb/consult--source-recentf-items)
```
I also implemented a unquified version. I've tried to hook into `uniquify.el`, but it seemed just too entangled with the buffer-list that I could see no way to do it. So I gave it a stab myself.

```elisp
(defun gb/consult--source-recentf-items-uniq ()
  (let ((ht (consult--buffer-file-hash))
        file-name-handler-alist ;; No Tramp slowdown please.
        items)
    (dolist (file (gb/recentf-list-uniq) (nreverse items))
      ;; Emacs 29 abbreviates file paths by default, see
      ;; `recentf-filename-handlers'.
      (unless (eq (aref (cdr file) 0) ?/)
        (setcdr file (expand-file-name (cdr file))))
      (unless (gethash (cdr file) ht)
        (push (propertize
               (car file)
               'multi-category `(file . ,(cdr file)))
              items)))))

(plist-put consult--source-recent-file
           :items #'gb/consult--source-recentf-items-uniq)

(defun gb/recentf-list-uniq ()
  (let* ((proposed (mapcar (lambda (f)
                             (cons (file-name-nondirectory f) f))
                           recentf-list))
         (recentf-uniq proposed)
         conflicts resol file)
    ;; collect conflicts
    (while proposed
      (setq file (pop proposed))
      (if (assoc (car file) conflicts)
          (push (cdr file) (cdr (assoc (car file) conflicts)))
        (if (assoc (car file) proposed)
            (push (list (car file) (cdr file)) conflicts))))
    ;; resolve conflicts
    (dolist (name conflicts)
      (let* ((files (mapcar (lambda (f)
                              ;; (file remaining-path curr-propos)
                              (list f
                                    (file-name-directory f)
                                    (file-name-nondirectory f)))
                            (cdr name)))
             (curr-step (mapcar (lambda (f)
                                  (file-name-nondirectory
                                   (directory-file-name (cadr f))))
                                files)))
        ;; Quick check, if there are no duplicates, we are done.
        (if (eq (length curr-step) (length (delete-dups curr-step)))
            (setq resol
                  (append resol
                          (mapcar (lambda (f)
                                    (cons (car f)
                                          (file-name-concat
                                           (file-name-nondirectory
                                            (directory-file-name (cadr f)))
                                           (file-name-nondirectory (car f)))))
                                  files)))
          (while files
            (let (files-remain)
              (dolist (file files)
                (let ((curr-propos (caddr file))
                      (curr-part (file-name-nondirectory
                                  (directory-file-name (cadr file))))
                      (rest-path (file-name-directory
                                  (directory-file-name (cadr file))))
                      (curr-step
                       (mapcar (lambda (f)
                                 (file-name-nondirectory
                                  (directory-file-name (cadr f))))
                               files)))
                  (if (member curr-part (cdr (member curr-part curr-step)))
                      ;; There is more than one curr-part in curr-step for
                      ;; this candidate.
                      (push (list (car file)
                                  rest-path
                                  (file-name-concat curr-part curr-propos))
                            files-remain)
                    ;; There is no repetition of curr-part in curr-step
                    ;; for this candidate.
                    (push (cons (car file)
                                (file-name-concat curr-part curr-propos))
                          resol))))
              (setq files files-remain))))))
    ;; apply resolved conflicts
    (let (items)
      (dolist (file recentf-uniq (nreverse items))
        (let ((curr-resol (assoc (cdr file) resol)))
          (if curr-resol
              (push (cons (cdr curr-resol) (cdr file)) items)
            (push file items)))))))
```

It is most likely not up to the standards of the house. It is also not on par with `uniquify.el` neither as to results (no exclusion of intermediary repeated path parts) nor as to implementation. Also, very lightly tested so far. But it what this non-programmer could get.  ;-)

I've tried though to scan for conflicts first, so that the most expensive parts only apply to conflicting candidates. It feels snappy here on a jurassic 7-8 year old computer, which was never quite powerful, and `recentf-max-saved-items` of 100. I'm happy with it. And if it's useful to someone else, happier.

**Comment by minad** [comment]
@gusbrs Thanks! Can you please add your code to the wiki with an explanation and link to this issue? I would appreciate that. Then it will be less likely that your code gets lost. Also I see the wiki as some staging ground for possible new additions. I've often heard that people find the wiki useful to learn about some additional Consult tweaks one could add to the Emacs config.

**Comment by gusbrs** [comment]
@minad Sure! :-) I don't have signed papers though, it is OK to add to the wiki despite that?

**Comment by minad** [comment]
@gusbrs Of course.

**Comment by gusbrs** [comment]
Done! If you have any amendments as to style of content, let me know.

**Comment by gusbrs** [comment]
For anyone interested, I've improved a little the function to uniquify the recent files candidates and have updated the wiki entry. Now things are closer to `uniquify.el` in that only the parts of the path that differ are kept to distinguish the candidates, making for shorter candidates. (I'm not sure results are identical to `uniquify` though).
